File: cc/binary_tree_preorder_traversal.cc
#include "base_include.h"
#include "tree_node.h"

class BinaryTreePreorderTraversal {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> res;
        for (TreeNode *curr = root; curr != nullptr; ) {
            if (nullptr == curr->left) {
                res.push_back(curr->val);
                curr = curr->right;
                continue;
            }
            TreeNode *post = curr->left;
            for ( ; post->right != nullptr && post->right != curr; post = post->right) {}
            if (nullptr == post->right) {
                res.push_back(curr->val);
                post->right = curr;
                curr = curr->left;
            } else {
                post->right = nullptr;
                curr = curr->right;
            }
        }
        return res;
    }
};

int main(int argc, char *argv[]) {
    return 0;
}

File: python/binary_tree_preorder_traversal.py
#!/usr/bin/env python3

from tree_node import TreeNode

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        res = list()
        curr = root
        while curr:
            if curr.left is None:
                res.append(curr.val)
                curr = curr.right
            else:
                post = curr.left
                while post.right and post.right != curr:
                    post = post.right
                if post.right == curr:
                    post.right = None
                    curr = curr.right
                else:
                    res.append(curr.val)
                    post.right = curr
                    curr = curr.left
        return res

def main():
    solver = Solution()

if __name__ == '__main__':
    main()
